### ISSUE TITLE
Responses-only SFT can save an all-NaN adapter after target truncation

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#532]
## Reproducer

I hit this while running a bounded Apple Silicon smoke train with:

- `Qwen/Qwen3.8-27B`
- text SFT, `train_on_responses_only: true`
- BF16 layer streaming
- LoRA r64 / alpha128
- batch 1, two Alpaca rows, one epoch
- `max_length: 64`
- PyTorch 2.12.1, Transformers 5.12.1, TRL 0.29.1, PEFT 0.20.0, MPS

The checkpoint's chat template inserts a system reasoning block. The two assistant spans begin at token positions 63 and 75. After truncation to 64, the first row has one supervised token and the second has zero.

Soup still runs both steps and exits 0 with `Training Complete`:

```
loss: 0
grad_norm: nan
entropy: nan
```

The saved adapter is not merely unchanged: all 105,906,176 LoRA values are NaN.

Raising `max_length` to 128 makes the same run healthy: 12 and 11 supervised labels, loss 2.2839 -> 1.8100, finite gradient norms, and every saved adapter value finite.

## Why this matters

A dataset can legitimately contain a long prompt whose assistant response is entirely truncated. One such row can poison the optimizer, but the run is reported as successful and publishes a dead adapter.

## Acceptance criteria

- Reject a post-tokenization training row when no causal-loss target remains after truncation.
- Name the row and `data.max_length` in the error.
- Add a control with one surviving assistant target and a failing zero-target row.
- Do not emit `Training Complete` or save a final adapter after a non-finite training state.
- Verify the guard fails if the zero-target check is removed.

I am taking this issue and will prepare a focused fix from current `main`.

**Comment by Amix29** [comment]
The focused fix is now available in #535. It rejects a post-tokenization row when `labels[1:]` contains no causal-loss target, reports the split, human-facing row number, and `data.max_length`, and refuses the final save if training metrics or trainable floating-point parameters are non-finite. The original Qwen3.8-27B reproducer is refused at `max_length=64` on train row 2 and accepted at `max_length=128` with 12 and 11 shifted targets. The full local suite passes: 18,794 passed, 82 skipped, 5 deselected.

**Comment by MakazhanAlpamys** [comment]
Shipped via #535 (@Amix29) - merged as aefc2cb.

SFT now refuses rows that would train on nothing: if tokenization/truncation leaves zero shifted causal-LM targets (`labels[1:]`) for a row, the run stops before Trainer construction with the failing split, human-facing row number and `data.max_length` named - instead of silently training an all-NaN adapter to exit 0. The final model save is additionally refused when logged metrics or trainable parameters are non-finite.

Try: a responses-only config with `max_length` shorter than the assistant span - it now fails fast naming `train row N`, rather than saving a dead artifact.
